power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -598,17 +598,130 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
     static_assert(((transformer_index_of<TransformerTypes> < sizeof...(TransformerTypes)) && ...));
 
+    struct BinarySearch {
+      public:
+        BinarySearch() = default;
+        BinarySearch(IntS tap_pos, IntS tap_min, IntS tap_max) { reset_bs(tap_pos, tap_min, tap_max); }
+
+        IntS get_bs_current_tap() const { return current; }
+        bool get_bs_last_down() const { return last_down; }
+        bool get_bs_inevitable_run() const { return inevitable_run; }
+        bool get_end_of_bs() const { return get_bs_tap_lower_bound() >= get_bs_tap_upper_bound(); }
+
+        void set_bs_current_tap(IntS _current_tap) { this->current = _current_tap; }
+        void set_bs_last_check(bool _last_check) { this->last_check = _last_check; }
+        void set_bs_inevitable_run(bool _inevitable_run) { this->inevitable_run = _inevitable_run; }
+
+        void recalibrate(bool strategy_max) {
+            if (get_bs_tap_reverse() ? strategy_max : !strategy_max) {
+                set_bs_tap_lower_bound(get_bs_current_tap());
+                set_bs_last_down(false);
+            } else {
+                set_bs_tap_upper_bound(get_bs_current_tap());
+                set_bs_last_down(true);
+            }
+        }
+
+        void propose_new_pos(bool strategy_max, bool above_range) {
+            auto is_down = above_range ? get_bs_tap_reverse() : !get_bs_tap_reverse();
+            if (get_bs_last_check()) {
+                set_bs_current_tap(is_down ? get_bs_tap_lower_bound() : get_bs_tap_upper_bound());
+                set_bs_inevitable_run(true);
+            } else {
+                set_bs_last_down(is_down);
+                adjust_bs(strategy_max);
+            }
+        }
+
+        IntS post_process(bool strategy_max, bool previous_down, bool& tap_changed) {
+            const bool prefer_higher = strategy_max ? !get_bs_tap_reverse() : get_bs_tap_reverse();
+            const auto tap_pos = search_bs(prefer_higher);
+            const auto tap_diff = tap_pos - get_bs_current_tap();
+            if (tap_diff == 0) {
+                tap_changed = false;
+                return tap_pos;
+            }
+            if ((tap_diff == 1 && previous_down) || (tap_diff == -1 && !previous_down)) {
+                set_bs_last_check(true);
+            }
+            tap_changed = true;
+            set_bs_current_tap(tap_pos);
+            return tap_pos;
+        }
+
+      private:
+        IntS lower_bound{-12};      // tap position lower bound
+        IntS upper_bound{12};       // tap position upper bound
+        IntS current{0};            // current tap position
+        bool last_down{false};      // last direction
+        bool last_check{false};     // last run checked
+        bool tap_reverse{false};    // tap range normal or reversed
+        bool inevitable_run{false}; // inevitable run
+
+        inline IntS get_bs_tap_lower_bound() const { return lower_bound; }
```

Comment:
here and below:
* `inline` is not necessary when it is declared in a class/struct. since the full function is resolvable at compile-time, you can make it `constexpr` (implies `inline`) or otherwise remove `inline` entirely
* the `bs` infix is not necessary because the class name already implies that it's `bs`. either be as generic as possible or as specific as necessary. (`duck.make_duck_sound()` vs `duck.make_sound()` vs `duck.quack()`). even more so when

Reply:
Everything is dynamic, there is nothing to `constexpr` here. Removing `bs_` is reasonable, but made the code less funny ;)

Reply:
dynamic or not is irrelevant. the address can be `constexpr` evaluated.

a `constexpr` function means that function represents an expression that **can** be evaluated at compile-time, not that it is mandatory to do so. Since a getter like this only returns a value, the compiler understands that it can read the value at any point in time by just replacing the function call with the address of the `lower_bound`. this enables pre-compiling instead of always merely inlining.

note that this is different from `consteval` in which case the actual contents are always evaluated at compile-time (`consteval` = constantly evaluated value, `constexpr` = a constant expression)

```cpp
struct A {
    auto get_a() const { return a_; }
    constexpr auto get_a_constexpr() const { return a_; }

    int a_{};
};

static_assert(A{}.get_a_constexpr() == 0); // OK
static_assert(A{1}.get_a_constexpr() == 1); // OK
// static_assert(A{}.get_a() == 0); // Not OK: compile error
// static_assert(A{1}.get_a() == 1); // Not OK: compile error
```

Reply:
But at compile time the dimension of bbinary_search_ is unknown ;)

Reply:
comment addressed in the latest commit.